lib/internal/fixed_queue.js

```diff
@@ -74,36 +84,48 @@ class FixedCircularBuffer {
   }
 
   push(data) {
-    this.list[this.top] = data;
-    this.top = (this.top + 1) & kMask;
+    const top = this.top;
+    this.list[top] = data;
+    this.top = (top + 1) & kMask;
   }
 
   shift() {
-    const nextItem = this.list[this.bottom];
+    const bottom = this.bottom;
+    const nextItem = this.list[bottom];
     if (nextItem === undefined)
       return null;
-    this.list[this.bottom] = undefined;
-    this.bottom = (this.bottom + 1) & kMask;
+    this.list[bottom] = undefined;
+    this.bottom = (bottom + 1) & kMask;
     return nextItem;
   }
 }
 
 module.exports = class FixedQueue {
   constructor() {
     this.head = this.tail = new FixedCircularBuffer();
+    this._spare = null;
   }
 
   isEmpty() {
     return this.head.isEmpty();
   }
 
   push(data) {
-    if (this.head.isFull()) {
-      // Head is full: Creates a new queue, sets the old queue's `.next` to it,
-      // and sets it as the new main queue.
-      this.head = this.head.next = new FixedCircularBuffer();
+    let head = this.head;
+    if (head.isFull()) {
+      // Head is full: reuse the spare buffer if available.
+      // Otherwise create a new one. Link it and advance head.
+      let nextBuf = this._spare;
+      if (nextBuf !== null) {
+        this._spare = null;
+        nextBuf.reset();
+      } else {
+        nextBuf = new FixedCircularBuffer();
+      }
+      head = head.next = nextBuf;
+      this.head = head;
```

Comment:
why not pushing it to the extreme :)?

```suggestion
      this.head = head = head.next = nextBuf;
```